ocaml/database/db_cache_types.ml

```diff
@@ -350,10 +350,10 @@ let remove_from_set key t =
   Schema.Value.Set (List.filter (fun x -> x <> key) t)
 
 exception Duplicate
-let add_to_map key value t =
+let add_to_map idempotent key value t =
   let t = Schema.Value.Unsafe_cast.pairs t in
-  if List.mem key (List.map fst t) then raise Duplicate;
-  Schema.Value.Pairs ((key, value) :: t)
+  if List.mem_assoc key t && (not idempotent || List.assoc key t <> value) then raise Duplicate;
```

Comment:
But that has wrong semantics too.
If `idempotent` then we shouldn't raise an exception when `List.assoc key t = value`,
and otherwise we should raise an exception if `List.mem_assoc key t` regardless of the value.